privacyidea/lib/eventhandler/usernotification.py

```diff
@@ -335,6 +355,17 @@ def do(self, action, options=None):
                     log.warning("Failed to send a notification email to user "
                                 "{0}".format(recipient))
 
+            elif action.lower() == "savefile":
+                spooldir = handler_options.get("spooldirectory", ".")
+                filename = handler_options.get("filename")
+                random = get_alphanum_str(16)
+                filename = filename.format(random=random, **tags)
+                try:
+                    with open(spooldir + "/" + filename, "w") as f:
```

Comment:
We should make sure that one cannot "break out" of ``spooldir`` by specifying a filename like ``../../foo/bar``.

Reply:
currently this works like a charm... But seriously, neither filename nor directory are user-controlled and if the admin defining the handler wants to do harm...
Anyway we should sanitize it in case it gets used for something else in the future.

Reply:
I agree that protecting against rogue admins isn't as urgent as protecting against rogue users :) But still, we should sanitize the filename here.
@cornelinux suggested that we define a root spool directory in ``pi.cfg`` (instead of the event handler configuration) and only allow filenames without ``..``. What do you think?

Reply:
You can use the following snipped to check the filename:
```python
# expand tags and remove possible leading '/'
filename = filename.format(random=random, **tags).lstrip(os.path.sep)
outfile = os.path.normpath(os.path.join(spooldir, filename))
if not outfile.startswith(spooldir):
    log.error('Cannot write outside of spooldir {0!s}!'.format(spooldir))
```

Reply:
I'll write a test to check for directory traversal.

Reply:
done in ac1c326d

Reply:
Nice